src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -881,17 +881,64 @@ public ImmutableArray<T> RemoveRange(ReadOnlySpan<T> items, IEqualityComparer<T>
                 return self.Remove(items[0], equalityComparer);
             }
 
-            var indicesToRemove = new SortedSet<int>();
-            foreach (var item in items)
+#nullable disable
+            var itemsDic = new Dictionary<T, int>(equalityComparer);
+#nullable restore
+            int nullValueCount = 0;
+            foreach (ref readonly T item in items)
             {
-                int index = -1;
-                do
+                if (item == null)
```

Comment:
I'm not sure if the null check is strictly necessary here. The original `self.IndexOf` doesn't seem to check for null inputs either so we might be able to skip this altogether without regression (a test verifying this would be great though!)

Reply:
Main reason is `Dictionary<,>` cannot support `null `key (throw `ArgumentNullException `if happen), and ImmutableArray's member methods (including .IndexOf()) needs to support `null `item.

Reply:
Is there existing test coverage for `null` items?

Reply:
Test added and find issue of current algorithm about `null`: because `Dictionary `cannot support `null` key and `equalityComparer `can make for example `null be equal to some value`, then even if there is no more `nullValueCount`, it still should use `Dictionary `and expect it to support null and respect the comparer. 

Imagine case:
```
items: [null, 1] 
ImmutableArray: [null, null]
equalityComparer: any value is same !
```

Then 2rd null of `ImmutableArray `should be able to be treated well in `Dictionary` but not.

Reply:
I'm not sure I fully understand. Are you saying that we should strip the `nullValueCount` and delegate `null` key handling to the dictionary, under the assumption that the equality comparer should handle null values? I don't think this works by default, although I think it can be forced by explicitly passing `equalityComparer ?? EqualityComparer<T>.Default` (IIRC `EqualityComparer<T>.Default` does handle null values correctly). 

The only potential source of regression here is having a user-defined equality comparer that throws when the value is null. Would it be possible to double check if that also happens in the case of the existing algorithm?

Reply:
I just added some demo used test code (and pushed it) to explain, we can check new test cases result to understand when CI completes (I had run it locally). 
I think the regression is new algorithm cannot handle (null & eg. `AlwaysEqualsOrAlwaysNotEqualsEqualityComparers`): it sound like our new code is putting null value into separated hash bucket (handling code) and put value `1` to another bucket (real hash bucket and cannot take `null`), but strange user (custom `equalityComparer`) wants them be same, so that our algorithm cannot meet it. Previous implementation has no this issue. 
Please tell if not clear.

```
[MemberData(nameof(RemoveRangeEnumerableNullData_WithAllEqualityComparer))]
case OldRemoveRangeCanSupportNullByAnyEqualityComparer can pass
```

```
[MemberData(nameof(RemoveRangeEnumerableNullData_WithDefaultEqualityComparer))]
case NewRemoveRangeEnumerableCanSupportNullByDefaultEqualityComparer can pass also
```

```
[MemberData(nameof(RemoveRangeEnumerableNullData_WithStrangeCustomEqualityComparer))]
case NewRemoveRangeEnumerableCannotSupportNullByStrangeCustomEqualityComparer will failed !
```

```
        public static IEnumerable<object[]> RemoveRangeEnumerableNullData_WithAllEqualityComparer()
        {
            return SharedEqualityComparers<int?>().SelectMany(comparer =>
                new[]
                {
                    new object[] { new int?[] { null, null }, new int?[] { null, 1 }, comparer }
                });
        }

        public static IEnumerable<object[]> RemoveRangeEnumerableNullData_WithDefaultEqualityComparer()
        {
            return DefaultEqualityComparers<int?>().SelectMany(comparer =>
                new[]
                {
                    new object[] { new int?[] { null, null }, new int?[] { null, 1 }, comparer }
                });
        }

        public static IEnumerable<object[]> RemoveRangeEnumerableNullData_WithStrangeCustomEqualityComparer()
        {
            return AlwaysEqualsOrAlwaysNotEqualsEqualityComparers<int?>().SelectMany(comparer =>
                new[]
                {
                    new object[] { new int?[] { null, null }, new int?[] { null, 1 }, comparer }
                });
        }
```

```
        private static IEnumerable<IEqualityComparer<T>> DefaultEqualityComparers<T>()
        {
            // Null comparers should be accepted and translated to the default comparer.
            yield return null;
            yield return EqualityComparer<T>.Default;
        }

        private static IEnumerable<IEqualityComparer<T>> AlwaysEqualsOrAlwaysNotEqualsEqualityComparers<T>()
        {
            yield return new DelegateEqualityComparer<T>(equals: (x, y) => true, objectGetHashCode: obj => 0, getHashCode: obj => 0);
            yield return new DelegateEqualityComparer<T>(equals: (x, y) => false, objectGetHashCode: obj => 0, getHashCode: obj => 0);
        }
```

Reply:
> it sound like our new code is putting null value into separated hash bucket

Right, so it's a concern about the special handling of `null` value (tracked independently in the `nullValueCount` variable) which creates issues for equality comparers that equate null values with non-null values?

I would recommend creating the dictionary using `new Dictionary<,>(equalityComparer ?? EqualityComparer<T>.Default)` which IIRC should correctly support `null` values without the special variable.

Reply:
> creating the dictionary using `new Dictionary<,>(equalityComparer ?? EqualityComparer<T>.Default)` which IIRC should correctly support `null` values without the special variable

Actually `Dictionary<,>` cannot support `null` key no matter equalityComparer: [related code](https://github.com/dotnet/runtime/blob/8c30c3366908429f1ece359de83fec665a8e02fb/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L503)

![image](https://user-images.githubusercontent.com/72254037/152906558-cbd59b67-dc90-49a8-9dd2-2c8c1ec72e96.png)

Reply:
Ah good point, I had forgotten we're hardcoding the null check in dictionary. In this case I think we should just leave the existing naive algorithm as-is, and we can investigate potential improvements in a separate PR. Thanks for investigating!

Reply:
Reverted.